R/model_functions.R

```diff
@@ -0,0 +1,307 @@
+
+#' Simple Ecosystem Model
+#' @param X = [leaf,wood,root,storage,som,SoilWater,stem density]
+#' @param params params
+#' @param timestep is in seconds, defaults to 30 min
+#' @param inputs: PAR, temp, VPD
+#' @param pest [phloem, xylem, leaf, root, stem] a vector of pest impacts, each entry in the vector
+#' represents the percent change in thepholme, yxlem, leaf, root, and stem based on the type of disruption.
+#' @author Michael C, Dietze <dietze@bu.edu>
+#' @return X
+#' @export
+SEM <- function(X, params, inputs, pest = c(0, 0, 0, 1, 0), p = 1800) {
+
+  # Check the parameter inputs
+  extra_params <- which(!names(params) %in% names(default_parameters))
+  msg_params   <- paste(names(params)[extra_params], collapse = ', ')
+  assertthat::assert_that(length(extra_params) == 0, msg = paste0('params contains unknown parameters: ', msg_params))
+  missing_params <- which(!names(default_parameters) %in% names(params))
+  msg_params <- paste(names(params)[missing_params], collapse = ', ')
+  assertthat::assert_that(length(missing_params) == 0, msg = paste0('params is missing the following: ', msg_params))
+
+  # Define constants used to conver from unit A -> unit B
+  ## convert umol/m2/sec -> Mg/ha/sec
+  k <- 1e-6 * 12 * 1e-6 * 10000 # mol/umol*gC/mol*Mg/g*m2/ha
+  ktree <- 1e-6 * 12 * 1e-3 # mol/umol*gC/mol*kg/g -> kg/m2/sec
+  kH2O <- 1e-6 * 18 * 10^-6 # mol/umol*gH2O/mol*m/g
+  P <- 101.325 ## average atm pressure (kPa)
+
+  ### Hydrology ###
+
+  ## evaporation (patch)
+  EVAP <- min(X[6] / timestep, 1.15 * params$gevap * (0.622 * inputs$VPD / P) / 1000) ## Bonan Eqn 13.10: rate = m/s
+  ## 1.15 kg/m3 = rho = density of air
+  ## 1000 at end converts kg/m2/s to m/s
+  ##**** CHECK UNITS (want m/s)
+  X[6] <- X[6] + inputs$precip / 1000 - EVAP * timestep
+
+  ## plant available water (trapazoidal response) (patch)
+  paw <- ifelse(X[6] > params$Wthresh, X[6] - 0.5 * params$Wthresh,
```

Comment:
I know this is Mike's code, not yours, but we should replace these vector-position numbers with named constants at some point.